### DISCUSSION TITLE
Namespacing

### DISCUSSION THREAD

**Discussion by repomaa** [luckyframework/lucky#770] [General]
This is a just a suggestions but I thing it would be much cleaner and less error prone if all generated code would be namespaced _at least_ to the application name. Even more preferable in my opinion would be further namespacing by `Models`, `Pages`, `Actions` etc.

**Comment by jwoertink** [comment]
I just ran in to 2 issues where namespacing would have helped. Working on an app for building out a computer with specs, and I wanted a model named `System`. That's a module in Crystal, so you can't have a class named `System`. The other is an app where you can plan to go to fun things where a model would be named `Event`, but since that exists in crystal, you can't use that name. 

Generating apps with a default namespace could help to solve these issues, I think. I'm not sure about the secondary level of namespace, but at least a top level would help.

**Reply by jkthorne** [reply]
I ran into the case with event and it threw me off.  It was very confusing.

**Reply by paulcsmith** [reply]
@wontruefree It looks like this has been fixed. When did you run into the error with `Event`? Was it awhile ago or recently?

**Comment by paulcsmith** [comment]
This is a good point. It would also helps with class collisions because you'd always need to preface it with the namespace `render Pages::Users::Index`

I have a few concerns/questions with this approach:

1. What if you want to namespace by a domain. Would it be `Admin::Actions::Users::Index` or `Actions::Admin::Users::Index`? I'm guessing the later (start with `Actions`) but is that strange? This would definitely make it harder to use "concept" based file structure if we decide to go that way https://github.com/luckyframework/lucky/issues/642
2. It is a bit extra to type. Probably not a big deal, but worth addressing. 
3. I'm not sure how this would work with models because queries and forms inherit from the model. `class Queries::User < Models::User::BaseQuery` seems a bit strange :S

Any other thoughts/concerns with this approach? Any thoughts about this @edwardloveall?

**Comment by paulcsmith** [comment]
Actually it looks like we can define top-level classes from within a class 🤯

```crystal
class User
  class ::User::BaseQuery
  end
end

User::BaseQuery.new # Works
```

So we could still have `class Queries::User < User::BaseQuery` but this seems a bit too magical ⚡️. Thoughts?

**Comment by jwoertink** [comment]
I'm not particularly keen on the secondary level namespacing idea, though I can see how it could be helpful. What I was seeing was something more like:

```crystal
module MyBlogApp
  class Admin::Users::Index < BrowserAction
  end
end

module MyBlogApp
  class UserQuery < User::BaseQuery
  end
end
```

In my use case, I could then do

```crystal
module FunApp
  class Event < BaseModel
    table :events do
    end
  end
end
```

This makes `FunApp::Event` an Avram Model, and `::Event` a crystal core class.

**Comment by paulcsmith** [comment]
Ah I see. One issue I have with this is that it makes guides/tutorials strange because some people use `MyApp` some use their real app name, etc. Also some apps names are super long! If we did do app based namespacing, I'd suggest we default to `App`. The guides would be simple, it's short, etc.

**Comment by jwoertink** [comment]
I'd be ok with a simple `module App` as the parent namespace. It would make docs consistent, and no guess work for generating stuff.

**Comment by z64** [comment]
I'm wholly unfamiliar with Lucky's codebase, but I read this issue and it sounded similar to a problem I had, so thought I would bring it up:

I don't know what would be appropriate for Lucky, but when I ran into this I solved it by using a simple marker annotation to collect classes at compile time that are meant to integrate with the app.

```cr
module MyWebsite
  @[Lucky::App]
  class Blog
    # ..
  end
end
```

Anyway, just an idea 🙂

**Comment by jwoertink** [comment]
I still don't understand annotations much. What would this do exactly? With lucky, you'd have something like

```crystal
class Event < Avram::Model
end
```

but since you can't have a class named `Event` in the global scope, we would need to namespace it. Could we use annotations to solve that?

**Comment by repomaa** [comment]
Hm any concluding thoughts? `module App` as parent namespace? Would you be willing to accept a PR that would implement this for the generators?

**Comment by jwoertink** [comment]
I've actually started using very custom namespaces because I have to share models between apps. I have 2 different models both named `Site` which connect to different databases, so they are each namespaced. My thought is this is probably a case by case scenario. 

If we do decide to add this in, then I think default should be no namespace, and then some flag or maybe just the syntax itself.

```
lucky gen.model User                # generates just class User
lucky gen.model App::User      # generates class App::User
lucky gen.model User -n App  # maybe I don't like this afterall... but same App::User
```

But I'll defer to @paulcsmith on this one.

**Comment by wout** [comment]
While not 100% related, I also ran into a namespace issue just now. One of the sections in my app is under `/settings`, so I have a `Settings::Index`. But this clashes with `Lucky::FormHelpers::Settings`. Explicitly defining it at the root (e.g. `::Settings::Index`) works, but it's not as nice.

**Reply by jwoertink** [reply]
Yeah, it's a whole mixed bag. On the one hand, it's tough dealing with all these name conflicts, but on the other hand, it would suck always having to type out a top level namespace for links and helpers like `link to: MyApp::Admin::Settings::Index`, and such.

I wonder if there's a way to guess at that stuff at compile time 🤔  Like if we could say "Hey, we see that you defined a class that has the same name as another already defined class/module/struct". It would be neat if crystal had a `macro redefined` or `macro reopened` like how you have inherited and included...

**Reply by wout** [reply]
Yeah, `link to: MyApp::Admin::Settings::Index` would not be an improvement. :joy:

But isn't this particular case easily solved with another naming convention for internal classes/structs? Or should `Lucky::FormHelpers::Settings` be easily accessible? I won't be the first one adding a `Settings` action.

**Reply by paulcsmith** [reply]
We should probably make Habitat define a private class for Settings if possible. Or HabitatSettings. I'll make an issue there for it.

Either way there is still the problem of namespace collisions like that :S I agree with Jeremy that it sucks when it happens, but I'm also nervous about having to type more for *every* action just to fix an edge case in *some* actions.

Maybe we can add something to lucky watch to look for these kinds of errors and suggesting adding `::` when it finds them...Ideally that could be something in Crystal though

**Reply by paulcsmith** [reply]
https://github.com/luckyframework/habitat/issues/45

**Reply by jwoertink** [reply]
Thanks for adding that issue. I think that's a good first step. As long as we make sure that Lucky is out of the way for stuff like this, I think it will for sure help. And Lucky supports using namespaces now, you just have to hand roll it (which isn't a bad solution).

**Reply by paulcsmith** [reply]
I also think that once we remove route inference there will be less collisions because you won't need to namespace actions with the parent resource (which was confusing anyway since you'd have actions in the parent folder that shouldn't be)

**Reply by wout** [reply]
Another option would be to add a warning at generator level for "reserved" names. But that doesn't help for manually generated classes, of course.

**Reply by paulcsmith** [reply]
@wout That is a great idea. We should also see if Crystal can namespace some of the common class names like `Event` to `Crystal::Event`

**Reply by jwoertink** [reply]
Those are actually already under the `Crystal` namespace. https://github.com/crystal-lang/crystal/blob/master/src/crystal/system/event_loop.cr#L23 I haven't tried in some time, but it's possible that the crystal namespace was just leaking out globally (i.e. `::Event` == `Crystal::Event`). Not totally sure. Maybe I will try making another app with those classes to see what happens. I should also note that when I originally tried, Lucky wasn't really a thing yet 😛

**Reply by paulcsmith** [reply]
Ohhh yeah. Just tried it. They must have fixed that already! 👍

**Reply by paulcsmith** [reply]
I like the idea of checking when generating a class, *but* so far I don't know of any. I knew Event was one, but looks like that was fixed. And `Settings` was from Lucky and we can fix that. So If someone runs into another collision I say we add some kind of check in the generator!

**Reply by wout** [reply]
Haven't seen any other collisions so far.
